content/blog/performant-code/index.Rmd

```diff
@@ -0,0 +1,526 @@
+median_time <- res$median[[1]]
+```
+
+There are two important lessons to take in from this output:
+
+* The `sum(unlist())` approach was wicked fast compared to `Reduce()`. 
+* Both of these expressions were fast. Even the slower of the two took `r format(median_time)`---to put that in perspective, that expression could complete `r format(round(1 / as.numeric(median_time)))` iterations in a second! Keeping this bigger picture in mind is always important when benchmarking; if code runs fast enough to not be an issue in practical situations, then it need not be optimized in favor of less readable or safe code.
+
+The results of little experiments like this one can be surprising at first. Over time, though, you will develop intuition for the fastest way to solve problems you commonly solve, and will write fast code the first time around! 
+
+### vctrs
+
+The problems I commonly solve---and possibly you as well, as a reader of this post---often involve lots of dplyr and tidyr. When profiling the tidymodels packages, I've come across many places where calls to dplyr and tidyr took more time than I'd like them to, but had a lot to learn about how to speed up those operations. _Enter the vctrs package!_
+
+```{r load-vctrs, message = FALSE, warning = FALSE}
+library(vctrs)
+```
+
+If you use dplyr and tidyr like I do, turns out you're also a vctrs user! dplyr and tidyr rely on vctrs to handle all sorts of elementary operations behind the scenes, and the package is a core part of a tidy developer's toolkit. Taken together with some functions from the tibble package, these tools provide a super efficient, albeit bare-bones, alternative interface to common data manipulation tasks like `filter()`ing and `select()`ing.
+
+## Rewriting tidy code
+
+For every performance improvement I make by rewriting dplyr and tidyr code to instead use vctrs and tibble, I make probably two or three simpler optimizations. [Tool-agnostic practices](https://adv-r.hadley.nz/perf-improve.html) such as reducing duplicated computations, implementing early returns where possible, and using vectorized implementations will likely take you far when optimizing R code. Profiling is your ground truth! When profiling indicates that otherwise well-factored code is slowed by tidy interfaces, though, all is not lost.
+
+We'll demonstrate different ways to speed up tidy code using a version of the base R data frame `mtcars` converted to a tibble:
+
+```{r mtcars-tbl}
+mtcars_tbl <- as_tibble(mtcars, rownames = "make_model")
+
+mtcars_tbl
+```
+
+### One-for-one replacements
+
+Many of the core functions in dplyr have alternatives in vctrs and tibble that can be quickly transitioned. There are a couple considerations associated with each, though, and some of them make piping a bit more awkward---most of the time, when I switch these out, I remove the pipe `%>%` as well.
+
+#### `filter()`
+
+The dplyr code:
+
+```{r filter, eval = FALSE}
+filter(mtcars_tbl, hp > 100)
```

Comment:
```suggestion
mtcars_tbl %>%
  filter(hp > 100)
```

Just because this is most likely what it will look like when you come across it in practice